netspeed/src/backend.c

```diff
@@ -327,3 +350,412 @@ get_wireless_info (DevInfo *devinfo)
+    }
+
+    return NL_SKIP;
+}
+
+static char *
+channel_width_name (enum nl80211_chan_width width)
+{
+    switch (width) {
+        case NL80211_CHAN_WIDTH_20_NOHT:
+            return _("20 MHz (no HT)");
+        case NL80211_CHAN_WIDTH_20:
+            return _("20 MHz");
+        case NL80211_CHAN_WIDTH_40:
+            return _("40 MHz");
+        case NL80211_CHAN_WIDTH_80:
+            return _("80 MHz");
+        case NL80211_CHAN_WIDTH_80P80:
+            return _("80+80 MHz");
+        case NL80211_CHAN_WIDTH_160:
+            return _("160 MHz");
+        case NL80211_CHAN_WIDTH_5:
+            return _("5 MHz");
+        case NL80211_CHAN_WIDTH_10:
+            return _("10 MHz");
+        default:
+            return _("unknown");
+    }
+}
+
+static int
+ieee80211_frequency_to_channel (int freq)
+{
+    /* see 802.11-2007 17.3.8.3.2 and Annex J */
+    if (freq == 2484)
+        return 14;
+    else if (freq < 2484)
+        return (freq - 2407) / 5;
+    else if (freq >= 4910 && freq <= 4980)
+        return (freq - 4000) / 5;
+    else if (freq <= 45000) /* DMG band lower limit */
+        return (freq - 5000) / 5;
+    else if (freq >= 58320 && freq <= 64800)
+        return (freq - 56160) / 2160;
+    else
+        return 0;
+}
+
+static int
+iface_cb (struct nl_msg *msg,
+          void          *arg)
+{
+    DevInfo *devinfo = (DevInfo*) arg;
+    struct genlmsghdr *gnlh = nlmsg_data (nlmsg_hdr (msg));
+    struct nlattr *tb_msg[NL80211_ATTR_MAX + 1];
+
+    nla_parse (tb_msg, NL80211_ATTR_MAX, genlmsg_attrdata (gnlh, 0),
+               genlmsg_attrlen(gnlh, 0), NULL);
+
+    if (tb_msg[NL80211_ATTR_MAC]) {
+        int len = MIN(ETH_ALEN, nla_len (tb_msg[NL80211_ATTR_MAC]));
+        memcpy (devinfo->hwaddr, nla_data (tb_msg[NL80211_ATTR_MAC]), len);
+    }
+
+    if (tb_msg[NL80211_ATTR_SSID]) {
+        char buf[G_MAXUINT8];
+        int len = nla_len (tb_msg[NL80211_ATTR_SSID]);
+        memcpy (buf, nla_data (tb_msg[NL80211_ATTR_SSID]), len);
+        buf [len] = '\0';
+        devinfo->essid = g_strescape (buf, NULL);
+        g_debug ("ssid: %s", buf);
+    }
+
+    if (tb_msg[NL80211_ATTR_WIPHY_FREQ]) {
+        uint32_t freq = nla_get_u32 (tb_msg[NL80211_ATTR_WIPHY_FREQ]);
+        char buf[100];
+        char *pos = buf;
+
+        pos += sprintf (pos, _("%d (%d MHz)"),
+                        ieee80211_frequency_to_channel (freq), freq);
+
+        if (tb_msg[NL80211_ATTR_CHANNEL_WIDTH])
+            sprintf (pos, _(", width: %s"),
+                     channel_width_name (nla_get_u32 (tb_msg[NL80211_ATTR_CHANNEL_WIDTH])));
+
+        devinfo->channel = g_strdup (buf);
+    }
+
+    return NL_SKIP;
+}
+
+void
+get_wireless_info (DevInfo *devinfo)
+{
+    struct nl80211_state nlstate = {.sock = NULL, .nl80211_id = -1};
+    int err, ret;
+    struct nl_msg *msg;
+    signed long long devidx = 0;
+
+    err = nl80211_init (&nlstate);
+    if (err || (nlstate.sock < 0)) {
```

Comment:
Now we got another build warning.
https://travis-ci.org/github/mate-desktop/mate-applets/jobs/707239813#L4724
```
backend.c: In function 'get_wireless_info':

backend.c:680:30: warning: ordered comparison of pointer with integer zero [-Wextra]

  680 |     if (err || (nlstate.sock < 0)) {

      |                              ^

In file included from /usr/lib64/glib-2.0/include/glibconfig.h:9,

                 from /usr/include/glib-2.0/glib/gtypes.h:32,

                 from /usr/include/glib-2.0/glib/galloca.h:32,

                 from /usr/include/glib-2.0/glib.h:30,

                 from /usr/include/gtk-3.0/gdk/gdkconfig.h:13,

                 from /usr/include/gtk-3.0/gdk/gdk.h:30,

                 from /usr/include/gtk-3.0/gtk/gtk.h:30,

                 from netspeed.c:27:

netspeed.c: In function 'update_quality_icon':

/usr/include/glib-2.0/glib/gmacros.h:811:63: warning: comparison of unsigned expression in '< 0' is always false [-Wtype-limits]
```

Reply:
done